README.md
# hakyll-favicon [![Build Status](https://travis-ci.org/elaye/mushu.svg?branch=master)](https://travis-ci.org/elaye/hakyll-favicon)

This library allows you to easily add favicons to your hakyll website.
You provide one SVG image and the library will convert it to different resolutions and generate the corresponding html.

## Dependencies

This library depends on `ImageMagick` to convert the images.

## Usage

First, add a `faviconsRules` that points to your main favicon in your `Site.hs` file like this:

```

import Hakyll.Favicon

main = hakyll $ do
  faviconsRules "images/favicon.svg" -- path to your favicon
  ...
```

Second, add a `favicons` field in your template `head`:

```
<head>
  ...
  $favicons$
</head>
```

Finally add the favicons context to your template context:

```
main = hakyll $ do
  ...
  match "index.html" $ do
    ...
    let ctx = ... `mappend`
              ... `mappend`
              faviconsField `mappend` -- add this
              defaultContext
    ...
```

The [example](example/) directory provides a minimal working example of this.

## Example

First, build the example:

    stack build

Then build the example page:

    stack exec example build
    
Or you can start a local server to serve the generated page:

    stack exec example watch

## Generated favicons

The following favicons are generated:

target       | format | sizes  | description
-------------|--------|--------|------------
all | .ico | 32, 64 | basic favicon
all | .png | 32 | basic favicon
iOS | .png | 144 | third-generation iPad with high-resolution Retina display
iOS | .png | 114 | iPhone with high-resolution Retina display
iOS | .png | 72 | first- and second-generation iPad
iOS, Android | .png | 57 | non-Retina iPhone, iPod Touch, and Android 2.1+ devices


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [0.1.0] - 2017-04-22


test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
import Hakyll
import Hakyll.Favicon (faviconsRules)

main :: IO ()
main = do

  let config = defaultConfiguration
                  { destinationDirectory = "test/out"
                  , providerDirectory = "test"
                  , storeDirectory = "test/out/_cache"
                  , inMemoryCache = True
                  }

  hakyllWith config $ do
    faviconsRules "favicon.svg"

  


src/Hakyll/Favicon.hs
module Hakyll.Favicon
( faviconsField
, faviconsRules
) where

import Data.Monoid ((<>))
import Data.List (intersperse)

import Hakyll
import System.FilePath (takeExtension, (</>))

import Debug.Trace (traceShow)

newtype IconSize = IconSize Int

instance Show IconSize where
  show (IconSize s) = show s

data IconType = Basic IconSize | Ico [IconSize] | IOS IconSize

data Favicon = Favicon IconType

favicons :: [Favicon]
favicons =
  [ Favicon (Ico [IconSize 32, IconSize 64])
  -- basic favicon
  , Favicon (Basic (IconSize 32))
  -- third-generation iPad with high-resolution Retina display
  , Favicon (IOS (IconSize 144))
  -- iPhone with high-resolution Retina display
  , Favicon (IOS (IconSize 114))
  -- first- and second-generation iPad
  , Favicon (IOS (IconSize 72))
  -- non-Retina iPhone, iPod Touch, and Android 2.1+ devices
  , Favicon (IOS (IconSize 57))
  ]

iosTemplate :: Compiler (Item String)
iosTemplate = makeItem "<link rel=\"apple-touch-icon-precomposed\" sizes=\"$size$x$size$\" href=\"$src$\">"

icoTemplate :: Compiler (Item String)
icoTemplate = makeItem "<link rel=\"shortcut icon\" href=\"favicon.ico\" type=\"image/x-icon\">"

basicTemplate :: Compiler (Item String)
basicTemplate = makeItem "<link rel=\"shortcut icon\" href=\"$src$\">"

faviconsField :: Context String
faviconsField = field "favicons" $ \_ -> do
  itemBody <$> faviconsCompiler favicons

faviconsCompiler :: [Favicon] -> Compiler (Item String)
faviconsCompiler favicons = do
  htmls <- mapM faviconCompiler favicons :: (Compiler [Item String])
  makeItem $ concatMap itemBody htmls

faviconCompiler :: Favicon -> Compiler (Item String)
faviconCompiler favicon@(Favicon faviconType) = case faviconType of
  Ico _ -> icoTemplate >>= applyAsTemplate ctx
  IOS size -> iosTemplate >>= applyAsTemplate (ctx <> constField "size" (show size))
  Basic _ -> basicTemplate >>= applyAsTemplate ctx
  where ctx = constField "src" (toUrl (faviconPath favicon))

faviconName :: Favicon -> String
faviconName (Favicon (Ico _)) = "favicon.ico"
faviconName (Favicon (IOS size)) = "favicon" ++ show size ++ ".png"
faviconName (Favicon (Basic size)) = "favicon" ++ show size ++ ".png"

faviconPath :: Favicon -> FilePath
faviconPath favicon@(Favicon (Ico _)) = faviconName favicon
faviconPath favicon = "images" </> "favicons" </> faviconName favicon
faviconPath favicon = "images" </> "favicons" </> faviconName favicon

faviconsRules :: Pattern -> Rules ()
faviconsRules ptn = match ptn $ mapM_ processFavicon favicons

processFavicon :: Favicon -> Rules ()
processFavicon favicon@(Favicon (Ico sizes)) = processIco favicon sizes
processFavicon favicon@(Favicon (IOS size)) = processPng favicon size
processFavicon favicon@(Favicon (Basic size)) = processPng favicon size

processIco :: Favicon -> [IconSize] -> Rules ()
processIco favicon sizes = version ("ico-" ++ concat (intersperse "-" (show <$> sizes))) $ do
  route $ customRoute $ \_ -> faviconPath favicon
  let
    cmd = "convert"
    args =
      [ "-background"
      , "none"
      , "svg:-"
      , "-define"
      , concat ["icon:auto-resize=", concat (intersperse "," (show <$> sizes))]
      , "+repage"
      , "ico:-"
      ]
  compile $ getResourceLBS >>= withItemBody (unixFilterLBS cmd args)

processPng :: Favicon -> IconSize -> Rules ()
processPng favicon (IconSize size) = version ("png" ++ show size) $ do
  route $ customRoute $ \_ -> faviconPath favicon
  let
    cmd = "convert"
    args =
      [ "-background"
      , "none"
      , "svg:-"
      , "-resize"
      , concat [show size, "x", show size, "!"]
      , "+repage"
      , "png:-"
      ]
  compile $ getResourceLBS >>= withItemBody (unixFilterLBS cmd args)
